ipa-core/src/protocol/basics/shard_fin.rs

```diff
@@ -0,0 +1,436 @@
+/// and vice versa. Decomposed view is used to perform additions,
+/// share is used to send data to other shards.
+#[derive(Debug, Default)]
+struct Histogram<HV: BooleanArray, const B: usize>
+where
+    Boolean: FieldSimd<B>,
+{
+    values: BitDecomposed<AdditiveShare<Boolean, B>>,
+    _marker: PhantomData<HV>,
+}
+
+impl<HV: BooleanArray, const B: usize> Histogram<HV, B>
+where
+    BitDecomposed<AdditiveShare<Boolean, B>>:
+        for<'a> TransposeFrom<&'a Vec<AdditiveShare<HV>>, Error = LengthError>,
+    Vec<AdditiveShare<HV>>:
+        for<'a> TransposeFrom<&'a BitDecomposed<AdditiveShare<Boolean, B>>, Error = LengthError>,
+    Boolean: FieldSimd<B>,
+{
+    pub fn new(input: &Vec<AdditiveShare<HV>>) -> Result<Self, LengthError> {
+        Ok(Self {
+            values: BitDecomposed::transposed_from(input)?,
+            _marker: PhantomData,
+        })
+    }
+
+    pub fn compose(&self) -> Vec<AdditiveShare<HV>> {
+        if self.values.is_empty() {
+            Vec::new()
+        } else {
+            // unwrap here is safe because we converted values from a vector during
+            // initialization, so it must have the value we need
+            Vec::transposed_from(&self.values).unwrap()
+        }
+    }
+}
+
+#[cfg(test)]
+impl<HV: BooleanArray, const B: usize> crate::test_fixture::Reconstruct<Vec<HV>>
+    for [Histogram<HV, B>; 3]
+where
+    Boolean: FieldSimd<B>,
+    BitDecomposed<AdditiveShare<Boolean, B>>:
+        for<'a> TransposeFrom<&'a Vec<AdditiveShare<HV>>, Error = LengthError>,
+    Vec<AdditiveShare<HV>>:
+        for<'a> TransposeFrom<&'a BitDecomposed<AdditiveShare<Boolean, B>>, Error = LengthError>,
+{
+    fn reconstruct(&self) -> Vec<HV> {
+        let shares = self.each_ref().map(Histogram::compose);
+        shares.reconstruct()
+    }
+}
+
+impl<C: ShardedContext, HV: BooleanArray, const B: usize> ShardAssembledResult<C>
+    for Histogram<HV, B>
+where
+    AdditiveShare<Boolean, B>: BooleanProtocols<C, B>,
+    Boolean: FieldSimd<B>,
+    // I mean... there must be a less-verbose way to write these bounds
+    Vec<AdditiveShare<HV>>:
+        for<'a> TransposeFrom<&'a BitDecomposed<AdditiveShare<Boolean, B>>, Error = LengthError>,
+    BitDecomposed<AdditiveShare<Boolean, B>>:
+        for<'a> TransposeFrom<&'a Vec<AdditiveShare<HV>>, Error = LengthError>,
+    <HV as Serializable>::Size: Add<Output: ArrayLength>,
+{
+    type SingleMessage = AdditiveShare<HV>;
+
+    fn empty() -> Self {
+        Self::default()
+    }
+
+    #[allow(clippy::manual_async_fn)]
+    fn merge<'a>(
+        &'a mut self,
+        ctx: C,
+        record_id: RecordId,
+        other: Self,
+    ) -> impl Future<Output = Result<(), Error>> + Send + 'a
+    where
+        C: 'a,
+    {
+        async move {
+            // todo: EightBit only works for 256 breakdowns. EightBitStep will panic if we try
+            // to add larger values
```

Comment:
this isn't the only place this is an issue. this works for now, but we might want to think about a design that binds the step size to the number of breakdowns.

Reply:
yep, it will panic if we add more than 8 bits, so it is not entirely out of control. But I agree - it is far less than ideal